segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorageTests.java

```diff
@@ -1317,6 +1317,101 @@ public void testBaseConcatWithDefragWithMinMaxLimits() throws Exception {
     }
 
+    /**
+     * Test Concat after repeated failure when concat using append mode is on.
+     *
+     * @throws Exception Exception if any.
+     */
+    @Test
+    public void testConcatUsingAppendsAfterWriteFailure() throws Exception {
+        long maxRollingSize = 20;
+        // Last chunk of target in these tests always has garbage at end which can not be overwritten.
+        testConcatUsingAppendsAfterWriteFailure(maxRollingSize,
+                new long[] {5},
+                new long[] {1, 2, 3, 4},
+                new int[] {},
+                new long[] {5, 10},
+                15);
+
+        // First chunk in source has garbage at the end.
+        testConcatUsingAppendsAfterWriteFailure(maxRollingSize,
+                new long[] {5},
+                new long[] {1, 2, 3, 4},
+                new int[]  {0}, // add garbage to these chunks
+                new long[] {5, 1, 9},
+                15);
+
+        // First two chunk in source has garbage at the end.
+        testConcatUsingAppendsAfterWriteFailure(maxRollingSize,
+                new long[] {5},
+                new long[] {1, 2, 3, 4},
+                new int[]  {0, 1},  // add garbage to these chunks
+                new long[] {5, 1, 2, 7},
+                15);
+
+        // First three chunks in source has garbage at the end.
+        testConcatUsingAppendsAfterWriteFailure(maxRollingSize,
+                new long[] {5},
+                new long[] {1, 2, 3, 4},
+                new int[]  {0, 1, 2},  // add garbage to these chunks
+                new long[] {5, 1, 2, 3, 4},
+                15);
+
+        // All chunks in source has garbage at the end.
+        testConcatUsingAppendsAfterWriteFailure(maxRollingSize,
+                new long[] {5},
+                new long[] {1, 2, 3, 4},
+                new int[]  {0, 1, 2, 3},  // add garbage to these chunks
+                new long[] {5, 1, 2, 3, 4},
+                15);
+    }
+
+    private void testConcatUsingAppendsAfterWriteFailure(long maxRollingSize,
+                                                         long[] targetLayoutBefore,
+                                                         long[] sourceLayout,
+                                                         int[] chunksWithGarbageIndex,
+                                                         long[] targetLayoutAfter,
+                                                         long expectedLength) throws Exception {
+        String targetSegmentName = "target";
+        String sourceSegmentName = "source";
+        SegmentRollingPolicy policy = new SegmentRollingPolicy(maxRollingSize); // Force rollover after every 20 byte.
+        ChunkedSegmentStorageConfig config = ChunkedSegmentStorageConfig.DEFAULT_CONFIG.toBuilder()
+                .maxSizeLimitForConcat(100)
+                .minSizeLimitForConcat(100)
+                .build();
+
+        TestContext testContext = getTestContext(config);
```

Comment:
Please modify `TestContext` to implement `AutoCloseable`, and close any resources it uses in the `close` method. Then update all usages of this class to clean up (decorate with `@Cleanup`).

Reply:
I made these changes in other PR, once that PR is merged and pulled into this PR I'll make this change.
Thanks

Reply:
Done.